keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -0,0 +1,524 @@
+
+What is required of the kube-apiserver is to be able to tell the KMS plugin which KEK (local KEK or KMS KEK) it should use to decrypt the incoming DEK. To do so, upon encryption, the KMS plugin could provide the encrypted local KEK as part of the `metadata` field in the `EncryptResponse`. The kube-apiserver would then store it in etcd next to the DEK. Upon decryption, the kube-apiserver provides the encrypted local KEK in `metadata` and `observed_key_id` from the last encryption when calling Decrypt. In case no encrypted local KEK is provided in the `metadata`, then we can assume key hierarchy is not used. The KMS plugin would query the external KMS to use the remote KEK to decrypt the DEK (same behavior as today). No state coordination is required between different instances of the KMS plugin.
+
+For the reference KMS plugin, the encrypted local KEK is stored in etcd via the `metadata` field, and once decrypted, it can be stored in memory as part of the KMS plugin cache to be used for encryption and decryption of DEKs. The encrypted local KEK is used as the key and the decrypted local KEK is stored as the value.
+
+```proto
+message EncryptResponse {
+    // The encrypted data.
+    bytes ciphertext = 1;
+    // The KMS key ID used for encryption operations.
+    // This can be used to drive rotation.
+    string current_key_id = 2;
+    // Additional metadata to be stored with the encrypted data.
+    // This metadata can contain the encrypted local KEK that was used to encrypt the DEK.
+    // Stored unencrypted in etcd.
+    map<string, string> metadata = 3;
+}
+```
+
+The `DecryptRequest` passes the same `current_key_id` and `metadata` returned by the previous `EncryptResponse` of this data as its `observed_key_id` and `metadata` for the decryption request.
+
+```proto
+message DecryptRequest {
+    // The data to be decrypted.
+    bytes ciphertext = 2;
+    // UID is a unique identifier for the request.
+    string uid = 3;
+    // The keyID that was provided to the apiserver during encryption.
+    // This represents the KMS KEK that was used to encrypt the data.
+    string observed_key_id = 4;
+    // Additional metadata that was sent by the KMS plugin during encryption.
+    map<string, string> metadata = 5;
+}
+
+message DecryptResponse {
+    // The decrypted data.
+    bytes plaintext = 1;
+    // The KMS key ID used to decrypt the data.
+    string current_key_id = 2;
+    // Additional metadata that was sent by the KMS plugin.
+    map<string, string> metadata = 3;
+}
+
+message EncryptRequest {
+    // The data to be encrypted.
+    bytes plaintext = 2;
+    // UID is a unique identifier for the request.
+    string uid = 3;
+}
+```
+
+In terms of storage, a new structured protobuf format is proposed. Similar to the proto serializer, it will use a magic number to detect when the stored data is in a format that it understands:
+
+```go
+encryptedProtoEncodingPrefix = []byte{'e', 'k', '8', 's', 0}
```

Comment:
This magic number looks not very magical. Is there a chance of this conflicting with something else? Should we add some random bytes to it?

Reply:
The default one used is https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/serializer/protobuf/protobuf.go#L36-L46. We're proposing a different magic number here. Do we still need to add more random bytes?